### PR TITLE
[SPARK-25104][SQL]Avro: Validate user specified output schema

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

With code changes in https://github.com/apache/spark/pull/21847 , Spark can write out to Avro file as per user provided output schema.

To make it more robust and user friendly, we should validate the Avro schema before tasks launched.

Also we should support output logical decimal type as BYTES (By default we output as FIXED)

## How was this patch tested?

Unit test

### CODE REVIEW

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala [hunk 1]
```diff
@@ -72,62 +73,70 @@ class AvroSerializer(rootCatalystType: DataType, rootAvroType: Schema, nullable:
...
+      case (IntegerType, INT) =>
         (getter, ordinal) => getter.getInt(ordinal)
-      case LongType =>
+      case (LongType, LONG) =>
         (getter, ordinal) => getter.getLong(ordinal)
-      case FloatType =>
+      case (FloatType, FLOAT) =>
         (getter, ordinal) => getter.getFloat(ordinal)
-      case DoubleType =>
+      case (DoubleType, DOUBLE) =>
```
Comment (dbtsai) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala]: Do we want to allow users to do casting up from catalystType to avroType? For example, catalystType float to avroType double.

**Comment by dbtsai** [comment]
LGTM.

**Review by dbtsai — APPROVED** [review]

File: external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala [hunk 2]
```diff
@@ -850,6 +851,62 @@ class AvroSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
     }
   }
 
+  test("throw exception if unable to write with user provided Avro schema") {
+    val input: Seq[(DataType, Schema.Type)] = Seq(
+      (NullType, NULL),
+      (BooleanType, BOOLEAN),
+      (ByteType, INT),
+      (IntegerType, INT),
```
Comment (dongjoon-hyun) [inline external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala]: Could you add `(ShortType, INT),`, too?

File: external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala [hunk 3]
```diff
@@ -850,6 +851,62 @@ class AvroSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
...
+      (LongType, LONG),
+      (FloatType, FLOAT),
+      (DoubleType, DOUBLE),
+      (BinaryType, BYTES),
+      (DateType, INT),
+      (TimestampType, LONG)
```
Comment (dongjoon-hyun) [inline external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala]: If the intention is to be exhaustive, `decimal` types? And, primitive to complex, and vice versa?

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala [hunk 4]
```diff
@@ -212,7 +222,7 @@ class AvroSerializer(rootCatalystType: DataType, rootAvroType: Schema, nullable:
   }
 
   private def resolveNullableType(avroType: Schema, nullable: Boolean): Schema = {
-    if (nullable) {
+    if (nullable && avroType.getType != NULL) {
```
Comment (gengliangwang) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala]: This fixes a trivial bug if `avroType` is NULL type.

File: external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala (hunk 3 above)
Comment (gengliangwang) [reply external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala]: @dongjoon-hyun Thanks, I have updated the test.

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala (hunk 1 above)
Comment (gengliangwang) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala]: Personally I would like to keep it simple as this PR propose. 
If data type casting needed, users can always do it in DataFrame before writing Avro files.

But if this is important, we can work on it.

Comment (dbtsai) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala]: Yeah, if someone feels it's important, let's do it in different PR.

**Comment by dbtsai** [comment]
Merged into master. Thanks.

**Pull request closed without merging**